### PR TITLE
Get rid of store_type/1 in auth backends

### PR DESCRIPTION
**Pull request by chrzaszcz**
### Before the change

Each auth backend had a `store_type/1` callback, which could return
'external', 'plain' or 'scram'.

There was a function called `store_type/1` in `ejabberd_auth` as well.
It traversed all backends, trying to determine the store type
for the whole host with a logic that e.g. favored 'external' over
'plain' (see the implementation for details).

Each SASL mechanism had its own password type, which was stored
in the sasl_mechanism ETS table on mechanism registration.
The possible password types were 'plain', 'digest', 'scram' and 'cert'.

To determine supported mechanisms in `cyrsasl:listmech/1`,
the store type for the current host was used to filter the SASL
mechanisms with the following rules:

```
Auth store_type     Supported password types
--------------      ------------------------
external            plain
scram               plain, scram, cert
plain               plain, digest, scram
```

### After the change

There is no notion of store types anymore - auth backends have a new
`supports_password_type/1` callback, which returns a boolean that
indicates the support for individual password types.

To determine support for a particular password type for the whole host,
the corresponding function in `ejabberd_auth` checks if any backend
supports it.

To determine supported mechanisms in `cyrsasl:listmech/1`,
for each mechanism it is enough to check if the current host supports
its password type.

Furthermore, the 'cert' password type is now supported only by the 'pki'
auth backend and no longer matches mechanisms supporting SCRAM.

Configuration is simplified,
e.g. the HTTP auth backend does not disable SASL EXTERNAL for the whole host.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[6293.1](https://travis-ci.org/esl/MongooseIM/jobs/515693869) / Erlang 19.3 / small_tests / d197d55359770cfbf079aa815d64e2ec180a5d31
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2254/6293/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2254/6293/small_tests.19.3/small/ct_run.mongooseim@localhost.2019-04-04_12.52.04/lib.mongooseim.logs/run.2019-04-04_12.52.04/suite.log.html)

**Comment by mongoose-im** [comment]
[6295.1](https://travis-ci.org/esl/MongooseIM/jobs/515705923) / Erlang 19.3 / small_tests / b09e9f34c049672daad03accf3f1acb9de7b4727
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2254/6295/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2254/6295/small_tests.19.3/small/ct_run.mongooseim@localhost.2019-04-04_13.18.29/lib.mongooseim.logs/run.2019-04-04_13.18.30/suite.log.html)

File: big_tests/tests/oauth_SUITE.erl [hunk 1]
```diff
@@ -360,9 +355,9 @@ extract_tokens(#xmlel{name = <<"iq">>, children = [#xmlel{name = <<"items">>} =
     RTD = exml_query:path(Items, [{element, <<"refresh_token">>}, cdata]),
     {base64:decode(ATD), base64:decode(RTD)}.
 
-get_auth_method() ->
+supports_password_type(PasswordType) ->
```
Comment (fenek) [inline big_tests/tests/oauth_SUITE.erl]: Not used in the suite actually.

**Comment by mongoose-im** [comment]
[6307.1](https://travis-ci.org/esl/MongooseIM/jobs/518692636) / Erlang 19.3 / small_tests / 332f0323ca896b4dcf5778d40a3ea82a30491da6
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2254/6307/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2254/6307/small_tests.19.3/small/ct_run.mongooseim@localhost.2019-04-11_10.00.08/lib.mongooseim.logs/run.2019-04-11_10.00.09/suite.log.html)

**Pull request merged**